### ISSUE TITLE
cross-compile: fipsinstall failures

### ISSUE THREAD

**Issue by dlegaultbbry** [openssl/openssl#18043]
This is on the latest released 3.0.2. Before enabling fips, I have the full library integrated and all tests pass on a QNX target. The software is cross-compiled on a host machine (Linux -> QNX) and then copied over to the target for testing. Basically, I copy the full source tree on the target as is to simplify things such as the framework seeing everything it needs.

I'm trying to get fips working on QNX and I'm htting the following error when running the 03-test_fipsinstall part.

original fips.cnf file before running tests

```
# cat test/fips.cnf 
openssl_conf = openssl_init

# Comment out the next line to ignore configuration errors
config_diagnostics = 1

.include fipsmodule.cnf

[openssl_init]
providers = provider_sect
alg_section = evp_properties

[evp_properties]
# Ensure FIPS non-approved algorithms in the FIPS module are suppressed (e.g.
# TEST-RAND). This also means that EVP_default_properties_is_fips_enabled()
# returns the expected value
default_properties = "fips=yes"

[provider_sect]
fips = fips_sect
```

Running the actual test (I truncated the steps after #3 because everything fails anyway after this one)

```
# HARNESS_VERBOSE=1 ./qnxtest.sh test_fipsinstall
OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
Warning: HARNESS_JOBS > 1 ignored with HARNESS_VERBOSE
00-prep_fipsmodule_cnf.t .. 
# The results of this test will end up in test-runs/prep_fipsmodule
1..1
HMAC : (Module_Integrity) : Pass
SHA1 : (KAT_Digest) : Pass
SHA2 : (KAT_Digest) : Pass
SHA3 : (KAT_Digest) : Pass
TDES : (KAT_Cipher) : Pass
AES_GCM : (KAT_Cipher) : Pass
AES_ECB_Decrypt : (KAT_Cipher) : Pass
RSA : (KAT_Signature) : RNG : (Continuous_RNG_Test) : Pass
Pass
ECDSA : (PCT_Signature) : Pass
ECDSA : (PCT_Signature) : Pass
DSA : (PCT_Signature) : Pass
TLS13_KDF_EXTRACT : (KAT_KDF) : Pass
TLS13_KDF_EXPAND : (KAT_KDF) : Pass
TLS12_PRF : (KAT_KDF) : Pass
PBKDF2 : (KAT_KDF) : Pass
SSHKDF : (KAT_KDF) : Pass
KBKDF : (KAT_KDF) : Pass
HKDF : (KAT_KDF) : Pass
SSKDF : (KAT_KDF) : Pass
X963KDF : (KAT_KDF) : Pass
X942KDF : (KAT_KDF) : Pass
HASH : (DRBG) : Pass
CTR : (DRBG) : Pass
HMAC : (DRBG) : Pass
DH : (KAT_KA) : Pass
ECDH : (KAT_KA) : Pass
RSA_Encrypt : (KAT_AsymmetricCipher) : Pass
RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
INSTALL PASSED
perl /data/openssl/build/nto-x86_64-o/util/wrap.pl ../../apps/openssl fipsinstall -module ../../providers/fips.so -provider_name fips -section_name fips_sect -out ../../test/fipsmodule.cnf => 0
ok 1 - fips install
ok
All tests successful.
Files=1, Tests=1,  6 wallclock secs ( 0.20 usr  0.01 sys +  3.83 cusr  0.04 csys =  4.08 CPU)
Result: PASS
03-test_fipsinstall.t .. 
# The results of this test will end up in test-runs/test_fipsinstall
1..29
fipsinstall: Use -help for summary.
perl /data/openssl/build/nto-x86_64-o/util/wrap.pl ../../apps/openssl fipsinstall -out fips.cnf -module -provider_name fips -macopt 'digest:SHA256' -macopt 'hexkey:00' -section_name fips_sect => 1
ok 1 - fipsinstall fail
Can't open "dummy.tmp" for reading, No such file or directory
01000000:error:80000002:system library:BIO_new_file:No such file or directory:crypto/bio/bss_file.c:67:calling fopen(dummy.tmp, r)
01000000:error:10000080:BIO routines:BIO_new_file:no such file:crypto/bio/bss_file.c:75:
VERIFY FAILED
perl /data/openssl/build/nto-x86_64-o/util/wrap.pl ../../apps/openssl fipsinstall -in dummy.tmp -module ../../providers/fips.so -provider_name fips -mac_name HMAC -macopt 'digest:SHA256' -macopt 'hexkey:00' -section_name fips_sect -verify => 1
ok 2 - fipsinstall verify fail
HMAC : (Module_Integrity) : Fail
Failed to load FIPS module
INSTALL FAILED
01000000:error:1C8000D5:Provider routines:SELF_TEST_post:missing config data:providers/fips/self_test.c:289:
01000000:error:1C8000E0:Provider routines:ossl_set_error_state:fips module entering error state:providers/fips/self_test.c:387:
01000000:error:1C8000D8:Provider routines:OSSL_provider_init_int:self test post failure:providers/fips/fipsprov.c:706:
01000000:error:078C0105:common libcrypto routines:provider_init:init fail:crypto/provider_core.c:910:name=fips
perl /data/openssl/build/nto-x86_64-o/util/wrap.pl ../../apps/openssl fipsinstall -out fips.cnf -module ../../providers/fips.so -provider_name fips -mac_name HMAC -macopt 'digest:SHA256' -macopt 'hexkey:00' -section_name fips_sect => 1
not ok 3 - fipsinstall
...
```

Contents of the fipsmodule.cnf which was successfully generated before the test_fipsinstall ran

```
# cat ../build/nto-x86_64-o/test/fipsmodule.cnf 
[fips_sect]
activate = 1
install-version = 1
conditional-errors = 1
security-checks = 1
module-mac = 67:87:4F:43:EB:89:E7:C2:BD:AA:9A:C5:17:90:29:CB:6A:57:95:DB:15:64:2F:B9:20:C5:56:20:C4:1C:0E:47
install-mac = 41:9C:38:C2:8F:59:09:43:2C:AA:2F:58:36:2D:D9:04:F9:6C:56:8B:09:E0:18:3A:2E:D6:CC:69:05:04:E1:11
install-status = INSTALL_SELF_TEST_KATS_RUN
```

I'm not sure where to look to find what could be the issue.

EDIT: After thinking about it some more, could it be that it's because there is no default openssl config file setup on the target? I would think that tests should not count on a system being properly configured/installed and should instead rely on being self sufficient in setup to achieve the testing goals.

My other qualm is the test is writing to a file which already exists in the test folder (fips.cnf). Perhaps it should be writing to a different file to avoid conflicts and make things easier to follow.

**Comment by t8m** [comment]
Do you have a `providers/fipsmodule.cnf` file in the build tree?

The weird thing is the 00-prep_fipsmodule.t passes so I do not see a reason why failing test should fail.

**Comment by t8m** [comment]
Yes, the `providers/fipsmodule.cnf` is not created by the `00-prep_fipsmodule.t` test.

Could you run manually the perl invocations from 00-prep_fipsmodule.t and the last failing one in the test_fipsinstall perhaps under strace to find out what is the difference?

Because as I said I do not see what is the difference between those two `openssl fipsinstall` invocations that should make the first pass but the other to fail.

**Comment by t8m** [comment]
Ah, now I understand it - the `hexkey:00` is wrong.

If you're not using the Makefile for tests, you need to copy the way how the tests are invoked from it. In `run_tests` target it also sets the FIPSKEY environment variable to the value that is defined in the Makefile. You need to copy that in your script.

**Comment by dlegaultbbry** [comment]
The way tests are run is like this which has always worked (even on 1.1.1) and it does work for all of them except that single one.

```
export SRCTOP=$(pwd);
export BLDTOP=$(pwd);
export HARNESS_JOBS=2;
EXE_SHELL="perl $(pwd)/util/wrap.pl" perl test/run_tests.pl ${1}
```

**Comment by t8m** [comment]
Would you please try exporting the FIPSKEY as I've said?

**Comment by t8m** [comment]
On 1.1.1 there was no FIPS support so no surprise that FIPSKEY was not there. :grin:

**Comment by t8m** [comment]
BTW, the error message is misleading - the important thing is the `HMAC : (Module_Integrity) : Fail` line.

**Comment by dlegaultbbry** [comment]
Thanks, it works with the FIPSKEY manual setup. Could you not change this and fetch the key in run_tests instead so that it would work with and without Makefile invocation for testing as that is not possible on embedded targets. In the meantime, I can add some manual steps to have the setup work by fetching the key in the Makefile before starting testing.

**Comment by t8m** [comment]
We could, the question then is why don't you use the Makefile when you're transferring it to the target system anyway?

**Comment by dlegaultbbry** [comment]
It's an embedded target, there's no make or any other toolchain infrastructure to execute the tests from the Makefile which is why we do it by calling the perl code directly. We copy the whole file tree over to the target because that's what the test framework expects to find the recipes, executables (apps/openssl, etc.), config files, etc.

It's not hard to grep the Makefile key value and export it before calling run_tests.pl. It's just another hoop non desktop systems have to handle. It would have been preferable for the test to completely fail if the FIPSKEY value was not set, this way it would have been obvious what the missing piece was.

**Comment by t8m** [comment]
Could you try if
```diff
diff --git a/test/recipes/03-test_fipsinstall.t b/test/recipes/03-test_fipsinstall.t
index d99974e467..79b45af795 100644
--- a/test/recipes/03-test_fipsinstall.t
+++ b/test/recipes/03-test_fipsinstall.t
@@ -27,7 +27,7 @@ plan skip_all => "Test only supported in a fips build" if disabled("fips");
 plan tests => 29;
 
 my $infile = bldtop_file('providers', platform->dso('fips'));
-my $fipskey = $ENV{FIPSKEY} // '00';
+my $fipskey = $ENV{FIPSKEY} // config('FIPSKEY') // '00';
 
 # Read in a text $infile and replace the regular expression in $srch with the
 # value in $repl and output to a new file $outfile.
```
work?

**Comment by t8m** [comment]
#18048

**Comment by BeyondWang** [comment]
> Yes that works. Thanks.

Hi， can 

> This is on the latest released 3.0.2. Before enabling fips, I have the full library integrated and all tests pass on a QNX target. The software is cross-compiled on a host machine (Linux -> QNX) and then copied over to the target for testing. Basically, I copy the full source tree on the target as is to simplify things such as the framework seeing everything it needs.
> 
> I'm trying to get fips working on QNX and I'm htting the following error when running the 03-test_fipsinstall part.
> 
> original fips.cnf file before running tests
> 
> ```
> # cat test/fips.cnf 
> openssl_conf = openssl_init
> 
> # Comment out the next line to ignore configuration errors
> config_diagnostics = 1
> 
> .include fipsmodule.cnf
> 
> [openssl_init]
> providers = provider_sect
> alg_section = evp_properties
> 
> [evp_properties]
> # Ensure FIPS non-approved algorithms in the FIPS module are suppressed (e.g.
> # TEST-RAND). This also means that EVP_default_properties_is_fips_enabled()
> # returns the expected value
> default_properties = "fips=yes"
> 
> [provider_sect]
> fips = fips_sect
> ```
> 
> Running the actual test (I truncated the steps after #3 because everything fails anyway after this one)
> 
> ```
> # HARNESS_VERBOSE=1 ./qnxtest.sh test_fipsinstall
> OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
> Warning: HARNESS_JOBS > 1 ignored with HARNESS_VERBOSE
> 00-prep_fipsmodule_cnf.t .. 
> # The results of this test will end up in test-runs/prep_fipsmodule
> 1..1
> HMAC : (Module_Integrity) : Pass
> SHA1 : (KAT_Digest) : Pass
> SHA2 : (KAT_Digest) : Pass
> SHA3 : (KAT_Digest) : Pass
> TDES : (KAT_Cipher) : Pass
> AES_GCM : (KAT_Cipher) : Pass
> AES_ECB_Decrypt : (KAT_Cipher) : Pass
> RSA : (KAT_Signature) : RNG : (Continuous_RNG_Test) : Pass
> Pass
> ECDSA : (PCT_Signature) : Pass
> ECDSA : (PCT_Signature) : Pass
> DSA : (PCT_Signature) : Pass
> TLS13_KDF_EXTRACT : (KAT_KDF) : Pass
> TLS13_KDF_EXPAND : (KAT_KDF) : Pass
> TLS12_PRF : (KAT_KDF) : Pass
> PBKDF2 : (KAT_KDF) : Pass
> SSHKDF : (KAT_KDF) : Pass
> KBKDF : (KAT_KDF) : Pass
> HKDF : (KAT_KDF) : Pass
> SSKDF : (KAT_KDF) : Pass
> X963KDF : (KAT_KDF) : Pass
> X942KDF : (KAT_KDF) : Pass
> HASH : (DRBG) : Pass
> CTR : (DRBG) : Pass
> HMAC : (DRBG) : Pass
> DH : (KAT_KA) : Pass
> ECDH : (KAT_KA) : Pass
> RSA_Encrypt : (KAT_AsymmetricCipher) : Pass
> RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
> RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
> INSTALL PASSED
> perl /data/openssl/build/nto-x86_64-o/util/wrap.pl ../../apps/openssl fipsinstall -module ../../providers/fips.so -provider_name fips -section_name fips_sect -out ../../test/fipsmodule.cnf => 0
> ok 1 - fips install
> ok
> All tests successful.
> Files=1, Tests=1,  6 wallclock secs ( 0.20 usr  0.01 sys +  3.83 cusr  0.04 csys =  4.08 CPU)
> Result: PASS
> 03-test_fipsinstall.t .. 
> # The results of this test will end up in test-runs/test_fipsinstall
> 1..29
> fipsinstall: Use -help for summary.
> perl /data/openssl/build/nto-x86_64-o/util/wrap.pl ../../apps/openssl fipsinstall -out fips.cnf -module -provider_name fips -macopt 'digest:SHA256' -macopt 'hexkey:00' -section_name fips_sect => 1
> ok 1 - fipsinstall fail
> Can't open "dummy.tmp" for reading, No such file or directory
> 01000000:error:80000002:system library:BIO_new_file:No such file or directory:crypto/bio/bss_file.c:67:calling fopen(dummy.tmp, r)
> 01000000:error:10000080:BIO routines:BIO_new_file:no such file:crypto/bio/bss_file.c:75:
> VERIFY FAILED
> perl /data/openssl/build/nto-x86_64-o/util/wrap.pl ../../apps/openssl fipsinstall -in dummy.tmp -module ../../providers/fips.so -provider_name fips -mac_name HMAC -macopt 'digest:SHA256' -macopt 'hexkey:00' -section_name fips_sect -verify => 1
> ok 2 - fipsinstall verify fail
> HMAC : (Module_Integrity) : Fail
> Failed to load FIPS module
> INSTALL FAILED
> 01000000:error:1C8000D5:Provider routines:SELF_TEST_post:missing config data:providers/fips/self_test.c:289:
> 01000000:error:1C8000E0:Provider routines:ossl_set_error_state:fips module entering error state:providers/fips/self_test.c:387:
> 01000000:error:1C8000D8:Provider routines:OSSL_provider_init_int:self test post failure:providers/fips/fipsprov.c:706:
> 01000000:error:078C0105:common libcrypto routines:provider_init:init fail:crypto/provider_core.c:910:name=fips
> perl /data/openssl/build/nto-x86_64-o/util/wrap.pl ../../apps/openssl fipsinstall -out fips.cnf -module ../../providers/fips.so -provider_name fips -mac_name HMAC -macopt 'digest:SHA256' -macopt 'hexkey:00' -section_name fips_sect => 1
> not ok 3 - fipsinstall
> ...
> ```
> 
> Contents of the fipsmodule.cnf which was successfully generated before the test_fipsinstall ran
> 
> ```
> # cat ../build/nto-x86_64-o/test/fipsmodule.cnf 
> [fips_sect]
> activate = 1
> install-version = 1
> conditional-errors = 1
> security-checks = 1
> module-mac = 67:87:4F:43:EB:89:E7:C2:BD:AA:9A:C5:17:90:29:CB:6A:57:95:DB:15:64:2F:B9:20:C5:56:20:C4:1C:0E:47
> install-mac = 41:9C:38:C2:8F:59:09:43:2C:AA:2F:58:36:2D:D9:04:F9:6C:56:8B:09:E0:18:3A:2E:D6:CC:69:05:04:E1:11
> install-status = INSTALL_SELF_TEST_KATS_RUN
> ```
> 
> I'm not sure where to look to find what could be the issue.
> 
> EDIT: After thinking about it some more, could it be that it's because there is no default openssl config file setup on the target? I would think that tests should not count on a system being properly configured/installed and should instead rely on being self sufficient in setup to achieve the testing goals.
> 
> My other qualm is the test is writing to a file which already exists in the test folder (fips.cnf). Perhaps it should be writing to a different file to avoid conflicts and make things easier to follow.

Hi, can you share how you cross-comile openssl for QNX? Thanks.

**Comment by shashikantjagtap7** [comment]
Hi, Can anyone tell me how to cross-compile the openssl for the QNX target?
